src/main/java/com/linkedin/kmf/services/ConsumeService.java

```diff
@@ -7,157 +7,148 @@
       Metrics metrics = new Metrics(metricConfig, reporters, new SystemTime());
       Map<String, String> tags = new HashMap<>();
-      tags.put(TAGS_NAME, _name);
-      _adminClient = AdminClient.create(props);
-      _sensors = new ConsumeMetrics(metrics, tags, topic, topicPartitionResult);
+      tags.put(TAGS_NAME, name);
+      _topic = consumerFactory.topic();
+      _sensors = new ConsumeMetrics(metrics, tags, _topic, topicPartitionResult, _adminClient, consumerFactory.latencyPercentileMaxMs(), consumerFactory.latencyPercentileGranularityMs());
+      _commitAvailabilityMetrics = new CommitAvailabilityMetrics(metrics, tags);
       _consumeThread = new Thread(() -> {
         try {
           consume();
         } catch (Exception e) {
-          LOG.error(_name + "/ConsumeService failed", e);
+          LOG.error(name + "/ConsumeService failed", e);
         }
-      }, _name + " consume-service");
+      }, name + " consume-service");
       _consumeThread.setDaemon(true);
-    });
+    }).get();
   }
 
   private void consume() throws Exception {
-    // Delay 1 second to reduce the chance that consumer creates topic before TopicManagementService
+    /* Delay 1 second to reduce the chance that consumer creates topic before TopicManagementService */
     Thread.sleep(1000);
 
     Map<Integer, Long> nextIndexes = new HashMap<>();
 
     while (_running.get()) {
       BaseConsumerRecord record;
       try {
-        record = _consumer.receive();
+        record = _baseConsumer.receive();
       } catch (Exception e) {
         _sensors._consumeError.record();
         LOG.warn(_name + "/ConsumeService failed to receive record", e);
-        // Avoid busy while loop
+        /* Avoid busy while loop */
         Thread.sleep(CONSUME_THREAD_SLEEP_MS);
         continue;
       }
 
-      if (record == null)
-        continue;
+      if (record == null) continue;
 
       GenericRecord avroRecord = Utils.genericRecordFromJson(record.value());
       if (avroRecord == null) {
         _sensors._consumeError.record();
         continue;
       }
       int partition = record.partition();
+
+      /* Commit availability and commit latency service */
+      try {
+        TopicPartition topicPartition = new TopicPartition(_topic, partition);
+        final AtomicReference<Exception> offsetCommitIssues = new AtomicReference<>(null);
+        OffsetAndMetadata offsetAndMetadata;
+        /* Call commitAsync, wait for a NON-NULL return value (see https://issues.apache.org/jira/browse/KAFKA-6183) */
+        OffsetCommitCallback commitCallback = (topicPartitionOffsetAndMetadataMap, exception) -> {
+          if (exception != null) {
+            offsetCommitIssues.set(exception);
+          }
+          _commitAvailabilityMetrics._offsetsCommitted.record();
+        };
+
+        long currTimeMillis = System.currentTimeMillis();
+
+        /* 5 seconds consumer offset commit interval. */
+        long timeDiffMillis = 5000;
```

Comment:
The more readable way of having time constants in the code is to do something like this:

```java
long timeDiffMillis = TimeUnit.SECONDS.toMillis(5);
```
@radai-rosenblatt  says there is another way using the Duration class that is more modern.

Reply:
Amended.